kvserver: use VoterConstraints when transferring leases

lunevalex opened the pull request:
TransferLeaseTarget and ShouldTransferLease use Constraints
to filter down the list of available stores to calculate the
mean leases across all the potential stores that could hold
a replica. With addition of non-voter replicas, we should be
pairing that list down to only the replicas that could hold voters.
This commit adds filtering by VoterConstraints to the lease
transfer logic.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63556)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-:-MYGOo-T9tytibn43t3P:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @lunevalex)

___
*[pkg/kv/kvserver/allocator.go, line 1445 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-MYGNXHO64iFF-o-WTjp:-MYGNXHO64iFF-o-WTjq:b132v8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcbeae8964ba97ba82d05464c09373ad3c45bef5/pkg/kv/kvserver/allocator.go#L1445)):*
> ```Go
> }
> 
> func (a Allocator) followTheWorkloadPrefersLocal(
> ```

Does the caller of this method, `StoreRebalancer.chooseLeaseToTransfer`, also need a similar treatment? It seems to only filter by overall constraints: `filteredStoreList := storeList.filter(zone.Constraints)`.
___
*[pkg/kv/kvserver/allocator_test.go, line 1360 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-MYGOB930VUmB_NkHsVv:-MYGOB930VUmB_NkHsVw:bbeuiww) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcbeae8964ba97ba82d05464c09373ad3c45bef5/pkg/kv/kvserver/allocator_test.go#L1360)):*
> ```Go
> 
> 	// 6 stores where the lease count for each store is equal to 10x the store
> 	// ID for odd stores and 0 for even.
> ```

Could you also describe how locality is distributed. I believe it's `dc=0` for even and `dc=1` for odd. It may just be worth writing out the full table:
```
1 | locality=dc=1 | lease_count=10
2 | locality=dc=0 | lease_count=0 
3 | locality=dc=1 | lease_count=30
4 | locality=dc=0 | lease_count=0
5 | locality=dc=1 | lease_count=50
6 | locality=dc=0 | lease_count=0
```


<!-- Sent from Reviewable.io -->

lunevalex commented:
___
*[pkg/kv/kvserver/allocator.go, line 1445 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-MYGNXHO64iFF-o-WTjp:-MYGdo5r1WY__fCOf00z:b-2yjir0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcbeae8964ba97ba82d05464c09373ad3c45bef5/pkg/kv/kvserver/allocator.go#L1445)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does the caller of this method, `StoreRebalancer.chooseLeaseToTransfer`, also need a similar treatment? It seems to only filter by overall constraints: `filteredStoreList := storeList.filter(zone.Constraints)`.
</blockquote></details>

Good catch, looking at this code it feels this filtering is happening too late, shouldn't it happen at the very beginning since we are using the storeList above to make decisions about QPS, even though arguably we should only be looking at QPS for the stores that meet the constraints? @aayushshah15 is this the thing that you are planning to fix? 


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> thanks for doing this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-:-MYHKyKx8Qqk0FTYG_g5:bxab46k)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @lunevalex and @nvanbenschoten)

___
*[pkg/kv/kvserver/allocator.go, line 1445 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-MYGNXHO64iFF-o-WTjp:-MYHIhC84nZaUNEbcBoc:b-v3wwdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcbeae8964ba97ba82d05464c09373ad3c45bef5/pkg/kv/kvserver/allocator.go#L1445)):*
What we seem to be doing _currently_ is that we're constraint -aware for rebalancing based on lease count (which is where the filtering inside `chooseLeaseToRebalance` comes into play)  but not based on QPS. It's a mess.

> is this the thing that you are planning to fix?

I think so, the solution to https://github.com/cockroachdb/cockroach/issues/61883 would have to fix this as well. 


<!-- Sent from Reviewable.io -->

Commit:
TransferLeaseTarget and ShouldTransferLease use Constraints
to filter down the list of available stores to calculate the
mean leases across all the potential stores that could hold
a replica. With addition of non-voter replicas, we should be
pairing that list down to only the replicas that could hold voters.
This commit adds filtering by VoterConstraints to the lease
transfer logic.

Release note: None

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63556#-:-MYLe_0w8auOwHNYeBOm:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @lunevalex)



<!-- Sent from Reviewable.io -->

lunevalex commented:
TFTR!

bors r+
